### PR TITLE
ELECTRON-978 (Implement native file picker modal window)

### PR DESCRIPTION
**Pull request by KiranNiranjan**
## Description
Exposes an API to the web app to open a native file picker modal [ELECTRON-978](https://perzoinc.atlassian.net/browse/ELECTRON-978)

## Solution Approach
Expose an API which invokes `dialog.showOpenDialog` inturn

## Screenshot
screenshot 2018-12-26 at 5 24 56 pm

## Related PRs
List related PRs against other branches / repositories:

branch | PR
------ | ------
SFE-Client-App | [#13157](https://github.com/SymphonyOSF/SFE-Client-App/pull/13157)
SFE-Packages | [#2239](https://github.com/SymphonyOSF/SFE-Packages/pull/2239)

## QA Checklist
- [X] Unit-Tests
- [ ] Automation-Tests

### CODE REVIEW

File: js/preload/preloadMain.js [hunk 1]
```diff
@@ -23,6 +23,9 @@ const { TitleBar } = require('../windowsTitlebar');
 const titleBar = new TitleBar();
 const { buildNumber } = require('../../package.json');
 const SnackBar = require('../snackBar').SnackBar;
+const fs = remote.require('fs');
```
Comment (VishwasShashidhar) [inline js/preload/preloadMain.js]: This is a security concern. We'll be exposing entire `fs` to preload right? Isn't there an alternative using `IPC`?

File: js/preload/preloadMain.js [hunk 2]
```diff
@@ -229,6 +232,36 @@ function createAPI() {
...
+         *  multiSelect: true,
+         *  filters: filters: [{ name: 'test', extensions: ['JPG', 'TXT', 'ai'] }]
+         * })
+         */
+        openFilePicker: function (options, callback) {
+            if (options && typeof callback === 'function') {
+                const browserWindow = remote.getCurrentWindow();
+                const { type, title, defaultPath, multiSelect, filters } = options;
+                const properties = [ type ];
+                if (multiSelect) properties.push('multiSelections');
```
Comment (VishwasShashidhar) [inline js/preload/preloadMain.js]: Please use braces around the if statement for consistency.

File: js/preload/preloadMain.js [hunk 3]
```diff
@@ -229,6 +232,36 @@ function createAPI() {
...
+                remote.dialog.showOpenDialog(browserWindow, { title, defaultPath, properties, filters }, (filePaths) => {
```
Comment (VishwasShashidhar) [inline js/preload/preloadMain.js]: Again, can we not use IPC here?

File: js/preload/preloadMain.js [hunk 4]
```diff
@@ -229,6 +232,36 @@ function createAPI() {
...
+                    if (!filePaths) return;
```
Comment (VishwasShashidhar) [inline js/preload/preloadMain.js]: Please use braces around the if statement for consistency.

File: js/preload/preloadMain.js [hunk 5]
```diff
@@ -229,6 +232,36 @@ function createAPI() {
...
+                    const files = filePaths.map((file) => {
+                        const mimeType = mime.lookup(file);
+                        const filename = file.substring(file.lastIndexOf('/') + 1);
+                        return new File([ fs.readFileSync(file) ], filename, { type: mimeType });
```
Comment (VishwasShashidhar) [inline js/preload/preloadMain.js]: If the file(s) is big & the system is slow, the UI thread will be blocked. Please use `async` version.

File: js/preload/preloadMain.js (hunk 1 above)
Comment (KiranNiranjan) [reply js/preload/preloadMain.js]: Currently, I don't think there are any solutions as IPC don't support buffers

File: js/preload/preloadMain.js (hunk 3 above)
Comment (KiranNiranjan) [reply js/preload/preloadMain.js]: Currently, I don't think there are any solutions

Comment (KiranNiranjan) [reply js/preload/preloadMain.js]: Currently, I don't think there are any solutions

**Comment by VishwasShashidhar** [comment]
We don't need this work around anymore as the issue has been fixed on Electron framework https://github.com/electron/electron/pull/16340

**Pull request closed without merging**